server/src/main/java/org/elasticsearch/search/lookup/LeafStoredFieldsLookup.java

```diff
@@ -37,14 +37,18 @@
 @SuppressWarnings({"unchecked", "rawtypes"})
 public class LeafStoredFieldsLookup implements Map<Object, Object> {
 
+    public interface FieldsReader {
```

Comment:
I was not able to follow why these changes are needed. Wouldn't everything keep working if context.reader() is provided as a constructor argument as before?

Reply:
This one is a pain - it's because some tests pass `null` as a context, and so we can't necessarily call `context.reader()`.  I initially changed it to just pass LeafReaderContext instead and then read by calling `context.reader().document()` but that falls foul of a test that was using mocking to build a fake Reader.  You can't mock LeafReaderContext because it's final.  Hence this refactoring - it does mean that we can remove the mocking though, which makes the test a lot easier to read.

Reply:
Curious what tests were failing because of a `null` context? To me it looks like the logic was just moved around, so I'm surprised we'd have a different in behavior (I think I'm missing something though...)

Reply:
It's the change in `FlattenedFieldLookupTests` from `LeafDocLookup docLookup = searchLookup.doc().getLeafDocLookup(null);` to `LeafDocLookup docLookup = searchLookup.getLeafSearchLookup(null).doc();`

Reply:
It seems to me that LeafReaderContext should never be null, and this specific test provides it null because it plugs in its own field data impl which makes the leaf reader context irrelevant. Then it sounds like we could simply find a way for this test to provide any non null value, rather than allowing for null in this way?

Would providing `Lucene.emptyReader(0).getContext()` work?

Reply:
Luca and I spoke offline about this, and agreed that the removal of the complicated mocking in FlattenedFieldLookupTests makes this change worth doing.

Reply:
Do you mean that we'll switch to creating a real context in `FlattenedFieldLookupTests` or keep this change? I don't have a strong opinion, and will approve so merging doesn't block on me. Last note: if we do end up sticking with the current strategy, maybe we could use `CheckedBiConsumer` to avoid introducing a new interface (with a very general name).

Reply:
We'll keep this change, and I have replaced the new interface with a CheckedBiConsumer.  Thanks for the reviews!